### DISCUSSION TITLE
BoostrapODPSample Not as Expected

### DISCUSSION THREAD

**Discussion by gracehymas** [casact/chainladder-python#408] [Q&A] [answered]
I am currently attempting to run a Bootstrap model on an incurred triangle, but the Ultimate results are not as expected. There is a massive amount of variability in the results, most noticeably so for the earlier Origin years. It does not make sense to me that OY 2002 would vary by even anything at all considering it is already fully developed in the triangle. I assume I must be calling a function wrong or perhaps there are some parameters I need to change. It seems that the BS and CL IBNR does converge well enough and using a different Excel based BS shows the expected trend. Hoping someone may be able to help I have researched this for days to no avail!

Code:
```
F_DF_t = cl.Triangle(DF_t, origin="YoA", development="Valuation", columns="Values", cumulative=True)
sims = cl.BootstrapODPSample(n_sims = 10000).fit(F_DF_t).resampled_triangles_
sim_ult = cl.Chainladder().fit(sims).ultimate_
```

Triangle:
![image](https://user-images.githubusercontent.com/56262552/207651585-ddba1614-cfd7-4bf3-9ee1-cfbccc767966.png)

Triangle (Flattened)
![image](https://user-images.githubusercontent.com/56262552/207651886-06507ce4-41b3-4549-8705-30511ee2d162.png)

Results:
![image](https://user-images.githubusercontent.com/56262552/207651235-97e04d76-f7cf-4e09-9e5b-6726c5dc6700.png)

**Comment by jbogaardt** [comment, marked as answer]
Hi @gracehymas , thanks for the question. 

A couple things to consider:
1. According to [CAS Monograph 4](https://www.casact.org/sites/default/files/2021-02/04-shapland.pdf), the Bootstrap model requires that incremental values be positive. This property is often violated by Incurred triangles. I believe (but haven't checked) the `BootstrapODPSample` throws these incrementals away which would lead to bias in the resampled triangles. They do offer a suggestion to get around this, but unfortunately, those suggestions are not yet implemented in `chainladder-python`. Excerpt from monograph:
![image](https://user-images.githubusercontent.com/19401133/207704617-69fb1b1b-9f77-43aa-9f39-dd09406107b5.png). I suspect this is causing some biases since your triangle does look like it would have negative incrementals.

2. As for the ultimates on OY2002 - these are the ultimates for each of the simulated triangles. They are not representative of the ultimate distribution for your original triangle. However, the simulated `ibnr_` reserves would appropriately translate to an `ibnr_` distribution on your original triangle. Here is an example looking at the IBNR distribution for the earliest origin year and indeed all values are zero.
```python
import chainladder as cl
tri = cl.load_sample('clrd')['CumPaidLoss'].groupby('LOB').sum().loc['wkcomp']
sims = cl.BootstrapODPSample().fit(tri).resampled_triangles_
cl_model = cl.Chainladder().fit(sims)
ibnr_dist = cl_model.ibnr_.iloc[:, :, 0].to_frame() # Earliest origin IBNR
ibnr_dist.max() == ibnr_dist.min() == 0
```

I suppose, following from the above example, you could translate this to an ultimate distribution by:
```python
ult_dist = tri.latest_diagonal + cl_model.ibnr_
```

**Reply by gracehymas** [reply]
Thank you so much for your thorough reply, I had considered that it could be negative increments but hoped the ODPBS would be able to cope! I may repeat the process with Paid Triangles instead to avoid this issue.

In terms of the ultimates, my aim was to use a percent rank to benchmark the output ultimates for each triangle - but your suggestion is also very useful and I will include it in my work!
